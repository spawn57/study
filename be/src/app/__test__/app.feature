Feature: Test Api
  Scenario: Hello Api
    Given a request
    When the api is triggered
    Then it matches the snapshot


 Scenario: Successful login
   Given a user with username "test" and password "test"
   When the api is triggered
   Then it matches the snapshot

