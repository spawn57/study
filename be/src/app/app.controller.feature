Feature: Login
 Scenario: Successful login
   Given a user with username "test" and password "test"
   When the user logs in with username "test" and password "test"
   Then the user should be logged in