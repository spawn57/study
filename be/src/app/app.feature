Feature: Login
 Scenario: Successful login
   Given a user with username "test" and password "test"
   When the api is triggered
   Then it matches the snapshot