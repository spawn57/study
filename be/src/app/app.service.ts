import { Injectable } from '@nestjs/common';

@Injectable()
export class AppService {
  getData(): { message: string } {
    return { message: 'Hello API' };
  }

  login(username: string, password: string): boolean {
    return true;
  }
}
