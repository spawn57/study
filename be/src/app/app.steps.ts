import { defineFeature, loadFeature } from 'jest-cucumber';

const feature = loadFeature(__dirname + `/app.feature`);

defineFeature(feature, (test) => {
  test('Successful login', ({ given, when, then }) => {
    let requestUrl; 
    let result;
    given(/^a user with username "(.*)" and password "(.*)"$/, (username, password) => { 
      requestUrl = '/login';
    });

    when(/the api is triggered/, () => {
      
    });

    then(/it matches the snapshot/, () => {
      expect(true).toMatchSnapshot();
    });
  });
});
