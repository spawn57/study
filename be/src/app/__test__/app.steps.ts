import { Test } from '@nestjs/testing';
import { defineFeature, loadFeature } from 'jest-cucumber';
import { AppModule } from '../app.module';
import { INestApplication } from '@nestjs/common';
import request from 'supertest';

const feature = loadFeature(__dirname + `/app.feature`);

defineFeature(feature, (test) => {
  test('Hello Api', ({ given, when, then }) => {
    let app: INestApplication;
    let req;
    let res;

    given(/^a request$/, async () => {
      const moduleRef = await Test.createTestingModule({
        imports: [AppModule],
      }).compile();

      req = {};

      app = moduleRef.createNestApplication();
      await app.init();
    });

    when(/the api is triggered/, async () => {
      res = await request(app.getHttpServer()).get('/').send(req);
    });

    then(/it matches the snapshot/, async () => {
      expect(res).toMatchSnapshot();
      await app.close();
    });
  });

  test('Successful login', ({ given, when, then }) => {
    let app: INestApplication;
    let req;
    let response;

    given(/^a user with username "(.*)" and password "(.*)"$/, async (username, password) => {
      const moduleRef = await Test.createTestingModule({
        imports: [AppModule],
      }).compile();

      req = {
        username: username,
        password: password,
      };
      app = moduleRef.createNestApplication();
      await app.init();
    });

    when(/the api is triggered/, async () => {
      response = await request(app.getHttpServer()).post('/login').send(req);
    });

    then(/it matches the snapshot/, async () => {
      expect(response).toMatchSnapshot();
      await app.close();
    });
  });
});
