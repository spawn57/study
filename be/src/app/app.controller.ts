import { Body, Controller, Get, Post, Req, Res } from '@nestjs/common';
import { AppService } from './app.service';
import { Request, Response } from 'express';

@Controller()
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get()
  getData() {
    return this.appService.getData();
  }

  @Post('/login')
  login(@Body() body: { username: string; password: string }, @Req() request: Request, @Res() response: Response) {
    const result = this.appService.login(body.username, body.password);
    response.sendStatus(200);
  }
}
