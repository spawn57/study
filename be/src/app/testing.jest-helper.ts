import { Test } from '@nestjs/testing';
import { AppModule } from './app.module';
import { INestApplication } from '@nestjs/common';
import { Response } from 'supertest';

export const setupApp = async () => {
  const moduleRef = await Test.createTestingModule({
    imports: [AppModule],
  }).compile();

  const app = moduleRef.createNestApplication();
  await app.init();

  return app;
};

export const dismantleApp = async (app: INestApplication) => {
  await app.close();
};


export const toMatchSnapshot =  (response: Response) => {
  const body = response.body;
  const text = response.text;

  const snapshot = {
    __status: response.status,
    ...(body && { body }),
    ...(text && { text }),
  }

  expect(snapshot).toMatchSnapshot();
}