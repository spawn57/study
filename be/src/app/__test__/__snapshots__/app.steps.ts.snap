// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Api Hello Api 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "23",
    "content-type": "application/json; charset=utf-8",
    "date": "Mon, 08 Jan 2024 12:55:34 GMT",
    "etag": "W/"17-TzGXketnjihp27LBaxXaC/GQ4Cw"",
    "x-powered-by": "Express",
  },
  "req": {
    "data": {},
    "headers": {
      "content-type": "application/json",
    },
    "method": "GET",
    "url": "http://127.0.0.1:33709/",
  },
  "status": 200,
  "text": "{"message":"Hello API"}",
}
`;

exports[`Test Api Successful login 1`] = `
{
  "header": {
    "connection": "close",
    "content-length": "2",
    "content-type": "text/plain; charset=utf-8",
    "date": "Mon, 08 Jan 2024 12:55:34 GMT",
    "etag": "W/"2-nOO9QiTIwXgNtWtBJezz8kv3SLc"",
    "x-powered-by": "Express",
  },
  "req": {
    "data": {
      "password": "test",
      "username": "test",
    },
    "headers": {
      "content-type": "application/json",
    },
    "method": "POST",
    "url": "http://127.0.0.1:43043/login",
  },
  "status": 200,
  "text": "OK",
}
`;
